### DISCUSSION TITLE
Image Factory  with Jetson Nano

### DISCUSSION THREAD

**Discussion by meitham** [siderolabs/talos#9221] [Q&A]
Hi Team, thanks for making this great distribution available.

I am trying to add a Jetson Nano to my cluster, and followed the docs on flashing it, but few things are not sure about.  The docs says use the "Vanilla" arm64 image, but my understanding that does not come with any Nvidia drivers, is that not the case?  Plugging my Jetson to HDMI screen I can see the Nvidia logo when the Jetson boots but the screen goes black afterward, though I can query with talosctl for disks on the nano.

The docs also say use interactive mode, but I don't see anything in my terminal when I run interactive, and the process goes forever until I have to kill it manually

`$ talosctl apply-config --insecure --mode=interactive --nodes jetson
^CSignal received, aborting, press Ctrl+C once again to abort immediately...`

I love the decalartive aspect of Talos but is it not possible to query the current machine configuration config for the jetson, modify it by hand and apply rather than doing interactive?  I could not find a way to retrieve machineconfig for nodes until they have already joined the cluster.

**Comment by smira** [comment]
I'm not exactly sure what is your question.

You can skip interactive installer if you wish, and follow [regular docs](https://www.talos.dev/v1.7/introduction/what-is-new/).

Create your [Jetson Nano image](https://factory.talos.dev/?platform=metal&target=sbc&version=1.7.6) via the Image Factory, you can NVIDIA extensions there as well.

**Comment by meitham** [comment]
Hi Smira, much appreciate coming to me, and apologies if my question isn't clear but because I have too many questions in two paragraphs.

# Flashing the firmware - 404
I have followed the docs for v1.7 reflashing, but the jetson release image provided [https://developer.nvidia.com/embedded/l4t/r32_release_v7.1/t210/jetson-210_linux_r32.7.2_aarch64.tbz2](https://developer.nvidia.com/embedded/l4t/r32_release_v7.1/t210/jetson-210_linux_r32.7.2_aarch64.tbz2) actually returns 404 so I ended up using [https://developer.nvidia.com/embedded/L4T/r32_Release_v4.4/r32_Release_v4.4-GMC3/T210/Tegra210_Linux_R32.4.4_aarch64.tbz2](https://developer.nvidia.com/embedded/L4T/r32_Release_v4.4/r32_Release_v4.4-GMC3/T210/Tegra210_Linux_R32.4.4_aarch64.tbz2) for the flash, and that appears to work ok, at least when booting with Jetpack (Ubuntu).

The `crane` command as part of the flash page refers to the OCI `ghcr.io/siderolabs/sbc-jetson:v0.1.0` but that doesn't exists yet, only `ghcr.io/siderolabs/sbc-jetson:v0.1.0-beta.1` exists as of today.

# Image Factory
Given the [docs page](https://www.talos.dev/v1.7/talos-guides/install/single-board-computers/jetson_nano/) is specefic to Nvidia Jetson I would expect the image provided there works out of the box (shows output on the screen via HDMI), but that's not the case.  The docs suggests using "Vanilla" Jetson Nano c7d6f36c6bdfb45fd63178b202a67cff0dd270262269c64886b43f76880ecf1e but does not mention if Nvidia GPU extensions are include by default or that I need to add them.  The question only arised because once Talos image boots into the unsafe mode, there is nothing on the screen.

# UART
After failing to get any output on the screen, I resorted to UART and connected the Jetson Nano to serial cable following [the provided guide](https://jetsonhacks.com/2019/04/19/jetson-nano-serial-console/) but again that worked with JetPack but not with Jetson.  Then I realised I needed a new  to download a new image from the factory with more customisation adding `console=ttyS0,115200` to the `extra kernel command` line arguments.  Unfortuantly that didn't work either and still nothing shows on my terminal via the serial port.

I have tried several combinations of images from the factory, such as ones with nvidia free and nvidia non-free.  Am I doing anything wrong here?

```
overlay:
    image: siderolabs/sbc-jetson
    name: jetson_nano
customization:
    extraKernelArgs:
        - console=ttyS0,115200
    systemExtensions:
        officialExtensions:
            - siderolabs/nonfree-kmod-nvidia
            - siderolabs/nvidia-container-toolkit
            - siderolabs/nvidia-open-gpu-kernel-modules
```

My cluster is currently on K3S and I am very keen to transition to Talos, and very happy to put my Jetson to test the pre-release if needed

**Comment by frezbo** [comment]
jetson nano is not going to support nvidia drivers, those drivers are only available for the nvidia provided sources and there's no oss or proprietary modules available to compile with talos. 

Seems nvidia have changed the link to https://developer.nvidia.com/embedded/l4t/r32_release_v7.2/t186/jetson_linux_r32.7.2_aarch64.tbz2

HDMI for jetson nano does not work, you'd need to use a serial console to debug

**Reply by meitham** [reply]
> HDMI for jetson nano does not work, you'd need to use a serial console to debug
Is this temporary or is there is a future plan to support HDMI?

Do I need to add any special customisation to to Talos to enable UART?

**Reply by frezbo** [reply]
the jetson image from talos should work, note that SBC's are community supported, so it could as well be not working as there is no maintainer for it now
